### PR TITLE
refactor(lint): modularize json linters

### PR DESCRIPTION
**Pull request by barrett-ruth**
refactor linters using json to use "from_json"

still wip

prep for "guard-collection"

### CODE REVIEW

**Comment by glepnir** [comment]
I think can `output_fmt` rename to `parser` ?

**Comment by barrett-ruth** [comment]
Ready for review.

Code actually works very nicely. All linters have been refactored - only 20 lines max!

Seeking advice in file `guard/lint.lua` - how to organize functions from_regex and from_json and not repeating code.

Should the file be extracted to `lua/guard/util.lua`? Or is it fine and wait for guard-collection before putting too much work in?

**Comment by glepnir** [comment]
The unified conversion interface looks good. But git log can be a bit confusing
Better to use git rebase.

File: lua/guard/lint.lua [hunk 1]
```diff
@@ -80,7 +80,126 @@ local function diag_fmt(buf, lnum, col, message, severity, source)
...
+    end, offences or {})
+
+    return diags
+  end
+end
+
+local regex_opts = {
+  regex = nil,
+  groups = { 'lnum', 'col', 'severity', 'code', 'message' },
+  -- TODO: DRY (watch out if need vim.deepcopy())
```
Comment (glepnir) [inline lua/guard/lint.lua:144]: can we find a best way to aviod use `deepcopy` in future ? Deep copy will cause some memory waste. and time consuming.

**Comment by barrett-ruth** [comment]
I'm bad at git lol. Closing this and re-opening with cleaner git history.

**Pull request closed without merging**